internal/mempool/mempool_abci.go

```diff
@@ -0,0 +1,260 @@
+//   return nil.
+//   TODO(berg): the above is not accurate, we do not error on that case. It makes
+//				 makes sense to include the peer state here in ABCI type and not the
+//				 mempool store itself.
+// - The transaction size exceeds the maximum transaction size as defined by the
+//   configuration provided to the mempool.
+// - The transaction fails Pre-Check (if it is defined).
+// - The proxyAppConn fails, e.g. the buffer is full.
+// NOTE:
+// - The applications' CheckTx implementation may panic.
+func (a *ABCI) CheckTx(ctx context.Context, tx types.Tx, callback func(*abci.ResponseCheckTx), txInfo TxInfo) error {
+	a.mtx.RLock()
+	defer a.mtx.RUnlock()
+
+	if txSize, maxTXBytes := len(tx), a.cfg.MaxTxBytes; txSize > maxTXBytes {
+		return types.ErrTxTooLarge{
+			Max:    maxTXBytes,
+			Actual: txSize,
+		}
+	}
+
+	if err := a.runPrecheck(tx); err != nil {
+		return types.ErrPreCheck{Reason: err}
+	}
+
+	if err := a.appConn.Error(); err != nil {
+		return err
+	}
+
+	// We add the transaction to the mempool's cache and if the
+	// transaction is already present in the cache, i.e. false is returned, then we
+	// check if we've seen this transaction and error if we have.
+	if !a.cache.Push(tx) {
+		// TODO(berg): the below is pulled directly from TxMempool. I don't understand what
+		// 			   it is doing? None of the returned types are being used? The comment
+		//			   feels before the if block feels misleading. We don't check if we've
+		//			   seen the tx before and error, we just call it and if it gets added
+		//			   to the txstore, so be it :shruggo:
+		//txmp.txStore.GetOrSetPeerByTxHash(txHash, info.SenderID)
+		return types.ErrTxInCache
+	}
+
+	res, err := a.appConn.CheckTx(ctx, abci.RequestCheckTx{Tx: tx})
+	if err != nil {
+		// TODO(berg): need something here for failed on CheckTx
+		return err
+	}
+
+	opRes, err := a.pool.CheckTXCallback(ctx, tx, res, txInfo)
+	if err != nil {
+		return err
+	}
+	a.applyOpResult(opRes)
+
+	if callback != nil {
+		callback(res)
+	}
+
+	return nil
+}
+
+// EnableTxsAvailable enables the mempool to trigger events when transactions
+// are available on a block by block basis.
+func (a *ABCI) EnableTxsAvailable() {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	a.txsAvailable = make(chan struct{}, 1)
+}
+
+// Flush clears the tx cache and the calls Flush on the underlying pool type.
+func (a *ABCI) Flush(ctx context.Context) error {
+	a.cache.Reset()
+	return a.pool.Flush(ctx)
+}
+
+// PoolMeta returns the metadata for the underlying pool implementation.
+func (a *ABCI) PoolMeta() PoolMeta {
+	return a.pool.Meta()
+}
+
+// PrepBlockFinality prepares the mempool for an upcoming block to be finalized. During
+// the execution of the new block, we want to make sure we are not running any additional
+// CheckTX calls to the application as well as flush any active connections underway.
+func (a *ABCI) PrepBlockFinality(ctx context.Context) (func(), error) {
```

Comment:
This is slightly different that the original `Lock/Unlock/FlushAppConn`, since the ABCI type controls the interations with the app, we can consolidate them into one function here.